pkg/i2gw/providers/gce/gce_extensions.go

```diff
@@ -0,0 +1,225 @@
+		notify(notifications.ErrorNotification, "No BackendConfig's found in annotation", service)
+		return "", false
+	}
+
+	if len(configs.Ports) != 0 {
+		notify(notifications.ErrorNotification, "Only config with default is supported since HealthCheckPolicy is attached on the whole service", service)
+		return "", false
+	}
+	return configs.Default, true
+}
+
+func ensureGceServiceIR(ctx context.Context, storage *storage, ir *i2gw.IR) {
+	if ir.Services == nil {
+		ir.Services = make(map[types.NamespacedName]*i2gw.ServiceIR)
+	}
+
+	beConfigToSvcs, betaBEConfigToSvcs := getBackendConfigMapping(ctx, storage)
+
+	for beConfigKey, beConfig := range storage.BackendConfigs {
+		gceServiceIR := beConfigToGceServiceIR(beConfig)
+		if gceServiceIR == nil {
+			continue
+		}
+
+		serviceList := beConfigToSvcs[beConfigKey]
+		for _, svcKey := range serviceList {
+			if ir.Services[svcKey] == nil {
+				ir.Services[svcKey] = &i2gw.ServiceIR{}
+			}
+			ir.Services[svcKey].Gce = gceServiceIR
+		}
+	}
+
+	for betaBeConfigKey, betaBeConfig := range storage.BetaBackendConfigs {
+		gceServiceIR := betaBeConfigToGceServiceIR(betaBeConfig)
+		if gceServiceIR == nil {
+			continue
+		}
+		serviceList := betaBEConfigToSvcs[betaBeConfigKey]
+		for _, svcKey := range serviceList {
+			if ir.Services[svcKey] == nil {
+				ir.Services[svcKey] = &i2gw.ServiceIR{}
+			}
+			ir.Services[svcKey].Gce = gceServiceIR
+		}
+	}
+}
+
+func beConfigToGceServiceIR(beConfig *backendconfigv1.BackendConfig) *i2gw.GceServiceIR {
+	if beConfig == nil {
+		return nil
+	}
+	var gceServiceIR i2gw.GceServiceIR
+	if beConfig.Spec.SessionAffinity != nil {
+		saConfig := i2gw.SessionAffinityConfig{
+			AffinityType: beConfig.Spec.SessionAffinity.AffinityType,
+			CookieTTLSec: beConfig.Spec.SessionAffinity.AffinityCookieTtlSec,
+		}
+		gceServiceIR.SessionAffinity = &saConfig
+	}
+
+	return &gceServiceIR
+}
+
+func betaBeConfigToGceServiceIR(betaBeConfig *backendconfigv1beta1.BackendConfig) *i2gw.GceServiceIR {
+	if betaBeConfig == nil {
+		return nil
+	}
+	var gceServiceIR i2gw.GceServiceIR
+	if betaBeConfig.Spec.SessionAffinity != nil {
+		saConfig := i2gw.SessionAffinityConfig{
+			AffinityType: betaBeConfig.Spec.SessionAffinity.AffinityType,
+			CookieTTLSec: betaBeConfig.Spec.SessionAffinity.AffinityCookieTtlSec,
+		}
+		gceServiceIR.SessionAffinity = &saConfig
+	}
+
+	return &gceServiceIR
+}
+
+func ensureGceServiceExtensions(ir i2gw.IR, gatewayResources *i2gw.GatewayResources) {
+	for svcKey, serviceIR := range ir.Services {
+		if serviceIR == nil {
+			continue
+		}
+		bePolicy := toGCPBackendPolicy(svcKey, serviceIR)
+		if bePolicy == nil {
+			continue
+		}
+		obj, err := i2gw.CastToUnstructured(bePolicy)
+		if err != nil {
+			notify(notifications.ErrorNotification, "Failed to cast GCPBackendPolicy to unstructured", bePolicy)
+			continue
+		}
+		key := types.NamespacedName{Namespace: obj.GetNamespace(), Name: obj.GetName()}
+		gatewayResources.GatewayExtensions[key] = *obj
```

Comment:
so you're treating a `GCPBackendPolicy` as a Gateway Extension here ? what happens when there are multiple kinds with the same namespaced name :) ?
if the goal is to use `GatewayExtensions` purely for marshaling reasons, then a list is better than a map

Reply:
+1

Reply:
Yes that is a valid concern. Changing it to list.